### ISSUE TITLE
Adding Multiple Tasks on a Counter only Increments Counter Once

### ISSUE THREAD

**Issue by cwfitzgerald** [RichieSams/FiberTaskingLib#61]
The following code fails

```
scheduler->AddTask(task1, &counter);
scheduler->AddTask(task2, &counter);

scheduler->WaitForCounter(&counter, 0);
```

Will cause issues because the counter is _set_ to one instead of being _incremented_ by one on https://github.com/RichieSams/FiberTaskingLib/blob/master/source/task_scheduler.cpp#L331. Similar problem on https://github.com/RichieSams/FiberTaskingLib/blob/master/source/task_scheduler.cpp#L354 with AddTasks. I didn't submit a PR because it's an easy enough fix and I have some other abstractions on my branch.

**Comment by RichieSams** [comment]
This change is ok. Just note that waiting on anything other than the "final" value of a counter is race-y. (Both before and after this change).

In addition, with this change, there is the possibility of logic-level undefined behavior. For exmaple:
```C++
AtomicCounter counter;

// store pointer to counter in some location that all threads can access
// ....

std::atomic_uint counter = 0;

Task task;
task.Function = [](ftl::TaskScheduler *taskScheduler, void *arg) {
	std::atomic_uint *counter = (std::atomic_uint *)arg;
	counter++;
};
task.Arg = &counter;


// Thread A                                     // Thread B
scheduler->AddTask(task, &counter);             scheduler->AddTask(task, &counter);
scheduler->WaitForCounter(&counter, 0);         scheduler->WaitForCounter(&counter, 0);

printf("%u", counter.load());                   printf("%u", counter.load());
```

The output is undefined. From a logic standpoint, you would expect:
```22```
However, if thread A happens to be OS scheduled before Thread B, and it's task is picked up immediately by a worker thread, Thread A's WaitForCounter will immediately return. Then B will run and do the same thing. This will result in:
```12```

**Comment by cwfitzgerald** [comment]
Yeah definitely. The uses I have for this paradigm or more on the order of adding multiple tasks from the same task, or communicating to parent tasks that continuations are still running. Example (abbreviated b/c of typing):

```
void funcA(...) {
    AtomicCounter counter(scheduler);
    blah struct_with_counter{&counter};
    scheduler->AddTask({funcB, struct_with_counter}, &counter); // counter is 1
    scheduler->WaitForCounter(&counter, 0); // Won't return until all continuations finish
}

void funcB(...) {
    // funcB is done, but I need a continuation...
    scheduler->AddTask({funcC, struct_with_counter}, struct_with_counter->counter); // Counter is 2
    scheduler->AddTask({funcD, struct_with_counter}, struct_with_counter->counter); // Counter is 3
} // Counter is 2

void funcC(...)
void funcD(...)
```

It's not very accurate, but you get the picture. I need A to wait for C and D without knowing they exist. Also my current abstraction only has a way to launch a single task at a time, so it needs the counter to work right with multiple tasks onto it.

**Comment by martty** [comment]
Fixed by b8a3b6118413744b8f2d2ad4c142d713add2d62a
